File: src/BrandVue.FrontEnd/Services/Exporter/CrosstabExporterUtilities/CrosstabHeader.cs
using BrandVue.Models;

namespace BrandVue.Services.CrosstabExporterUtilities
{
    public class CrosstabHeader
    {
        public string Id { get; }
        public string Name { get; }
        public char SignificanceIdentifier { get; }
        public CrosstabHeader[] SubHeaders { get; }
        public int Depth { get; }
        public int ColumnSpan { get; }

        private CrosstabHeader(string id, string name, char significanceIdentifier, CrosstabHeader[] subHeaders)
        {
            Id = id;
            Name = name;
            SignificanceIdentifier = significanceIdentifier;
            SubHeaders = subHeaders;
            Depth = CalculateDepth();
            ColumnSpan = CalculateColumnSpan();
        }

        public CrosstabHeader(CrosstabCategory category) : this(category.Id, category.DisplayName ?? category.Name, category.SignificanceIdentifier, category.SubCategories.Select(c => new CrosstabHeader(c)).ToArray()) { }

        public CrosstabHeader ExtendToDepth(int depth)
        {
            if (Depth == depth)
            {
                return this;
            }
            var extended = new CrosstabHeader(Id, null, default, new[] { this });
            return extended.ExtendToDepth(depth);
        }

        public CrosstabHeader[] GetColumnsAtDepth(int depth)
        {
            if (Depth == depth)
            {
                return new[] { this };
            }
            return SubHeaders.SelectMany(h => h.GetColumnsAtDepth(depth)).ToArray();
        }

        private int CalculateDepth()
        {
            if (SubHeaders.Length == 0)
            {
                return 0;
            }

            return 1 + SubHeaders.Select(s => s.Depth).Max();
        }

        private int CalculateColumnSpan()
        {
            if (SubHeaders.Length == 0)
            {
                return 1;
            }
            return SubHeaders.Select(s => s.ColumnSpan).Sum();
        }
    }
}

File: src/BrandVue.FrontEnd/client/components/visualisations/Crosstab/CrosstabHeader.ts
import { CrosstabCategory } from "../../../BrandVueApi";

export class CrosstabHeader {
    id: string;
    name: string | undefined;
    significanceIdentifier: string | undefined;
    subHeaders: CrosstabHeader[];
    depth: number;
    columnSpan: number;
    columnHasData: boolean;

    public constructor(id: string, name: string | undefined, significanceIdentifier: string | undefined, subHeaders: CrosstabHeader[]) {
        this.id = id;
        this.name = name;
        this.significanceIdentifier = significanceIdentifier;
        this.subHeaders = subHeaders;
        this.depth = this.getDepth(subHeaders);
        this.columnSpan = this.getColumnSpan(subHeaders);
    }

    public static fromApi(category: CrosstabCategory): CrosstabHeader {
        return new CrosstabHeader(category.id,
            category.displayName ?? category.name,
            category.significanceIdentifier,
            category.subCategories.map(CrosstabHeader.fromApi))
    }

    public extendToDepth(depth: number): CrosstabHeader {
        if (this.depth === depth) return this;

        const newRoot = new CrosstabHeader(this.id, undefined, undefined, [this]);
        return newRoot.extendToDepth(depth);
    }

    public getColumnsAtDepth(depth: number): CrosstabHeader[] {
        if (this.depth === depth) return [this];
        return this.subHeaders.reduce<CrosstabHeader[]>((all, subHeader) => all.concat(subHeader.getColumnsAtDepth(depth)), [])
    }

    private getDepth(subHeaders: CrosstabHeader[]): number {
        if (subHeaders.length === 0) return 0;

        return 1 + Math.max(...subHeaders.map(s => s.depth));
    }

    private getColumnSpan(subHeaders: CrosstabHeader[]): number {
        if (subHeaders.length === 0) return 1;
        return subHeaders.reduce((sum, subHeader) => sum + subHeader.columnSpan, 0);
    }
}